File: frameworks/hanami/app/actions/echo.rb
module Arena
  module Actions
    class Echo < Arena::Action
      def handle(request, response)
        # Read to end regardless of framing, so a chunked request works
        # without a Content-Length to size it from.
        chunks = []
        buf = request.body

        while (chunk = buf.read(65_536))
          chunks << chunk
        end

        response.headers["Content-Type"] = "application/octet-stream"
        response.body = chunks.join
      end
    end
  end
end

File: frameworks/effinitive/app/Tests/Echo.cs
using EffinitiveFramework.Core;
using EffinitiveFramework.Core.Http;

namespace effinitive.Tests;

public class EchoEndpoint : NoRequestEndpointBase<byte[]>
{
    protected override string Method => "POST";
    protected override string Route => "/echo";
    protected override string ContentType => "application/octet-stream";

    // The bytes that arrived go back unchanged. ReadBodyAsync reads to end
    // regardless of framing, so a chunked request works without a
    // Content-Length to size it from.
    public override async ValueTask<byte[]> HandleAsync(CancellationToken ct)
        => (await HttpContext!.ReadBodyAsync(ct)).ToArray();
}
